File: PawNClaw.Backend/PawNClaw.Data/Database/PetCenter.cs
﻿using System;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.EntityFrameworkCore;

#nullable disable

namespace PawNClaw.Data.Database
{
    [Index(nameof(Id), Name = "i")]
    public partial class PetCenter
    {
        public PetCenter()
        {
            Bookings = new HashSet<Booking>();
            CageTypes = new HashSet<CageType>();
            Cages = new HashSet<Cage>();
            CancelLogs = new HashSet<CancelLog>();
            CustomerVoucherLogs = new HashSet<CustomerVoucherLog>();
            GeneralLedgers = new HashSet<GeneralLedger>();
            Services = new HashSet<Service>();
            Supplies = new HashSet<Supply>();
            Vouchers = new HashSet<Voucher>();
            staff = new HashSet<Staff>();
        }

        [Key]
        [Column("id")]
        public int Id { get; set; }
        [Required]
        [Column("name")]
        [StringLength(256)]
        public string Name { get; set; }
        [Column("address")]
        [StringLength(256)]
        public string Address { get; set; }
        [Column("phone")]
        [StringLength(32)]
        public string Phone { get; set; }
        [Column("rating", TypeName = "numeric(19, 5)")]
        public decimal? Rating { get; set; }
        [Column("create_date", TypeName = "date")]
        public DateTime? CreateDate { get; set; }
        [Column("modify_date", TypeName = "date")]
        public DateTime? ModifyDate { get; set; }
        [Column("create_user")]
        public int? CreateUser { get; set; }
        [Column("modify_user")]
        public int? ModifyUser { get; set; }
        [Column("status")]
        public bool? Status { get; set; }
        [Column("brand_id")]
        public int BrandId { get; set; }
        [Column("open_time")]
        [StringLength(32)]
        public string OpenTime { get; set; }
        [Column("close_time")]
        [StringLength(32)]
        public string CloseTime { get; set; }
        [Column("description")]
        [StringLength(512)]
        public string Description { get; set; }
        [Column("checkin")]
        [StringLength(256)]
        public string Checkin { get; set; }
        [Column("checkout")]
        [StringLength(256)]
        public string Checkout { get; set; }

        [NotMapped]
        public int RatingCount { get => _getRatingCount(this.Bookings); }

        [NotMapped]
        public DateTime EndBooking { get; set; }

        [NotMapped]
        public DateTime OpenTimeDate { get; set; }

        [NotMapped]
        public DateTime CloseTimeDate { get; set; }

        [NotMapped]
        public DateTime CheckinDate { get; set; }

        [NotMapped]
        public DateTime CheckoutDate { get; set; }

        [ForeignKey(nameof(BrandId))]
        [InverseProperty("PetCenters")]
        public virtual Brand Brand { get; set; }
        [ForeignKey(nameof(CreateUser))]
        [InverseProperty(nameof(Account.PetCenterCreateUserNavigations))]
        public virtual Account CreateUserNavigation { get; set; }
        [ForeignKey(nameof(ModifyUser))]
        [InverseProperty(nameof(Account.PetCenterModifyUserNavigations))]
        public virtual Account ModifyUserNavigation { get; set; }
        [InverseProperty("IdNavigation")]
        public virtual Location Location { get; set; }
        [InverseProperty(nameof(Booking.Center))]
        public virtual ICollection<Booking> Bookings { get; set; }
        [InverseProperty(nameof(CageType.Center))]
        public virtual ICollection<CageType> CageTypes { get; set; }
        [InverseProperty(nameof(Cage.Center))]
        public virtual ICollection<Cage> Cages { get; set; }
        [InverseProperty(nameof(CancelLog.Center))]
        public virtual ICollection<CancelLog> CancelLogs { get; set; }
        [InverseProperty(nameof(CustomerVoucherLog.Center))]
        public virtual ICollection<CustomerVoucherLog> CustomerVoucherLogs { get; set; }
        [InverseProperty(nameof(GeneralLedger.Center))]
        public virtual ICollection<GeneralLedger> GeneralLedgers { get; set; }
        [InverseProperty(nameof(Service.Center))]
        public virtual ICollection<Service> Services { get; set; }
        [InverseProperty(nameof(Supply.Center))]
        public virtual ICollection<Supply> Supplies { get; set; }
        [InverseProperty(nameof(Voucher.Center))]
        public virtual ICollection<Voucher> Vouchers { get; set; }
        [InverseProperty(nameof(Staff.Center))]
        public virtual ICollection<Staff> staff { get; set; }

        [NotMapped]
        public ICollection<Photo> Photos { get; set; }

        [NotMapped]
        public decimal RatingPoint { get => _getRatingPoint(this.Bookings); }

        private int _getRatingCount(ICollection<Booking> Bookings)
        {
            int count = 0;
            foreach (var booking in Bookings)
            {
                if (booking.Rating.HasValue)
                {
                    count++;
                }
            }
            return count;
        }
        private decimal _getRatingPoint(ICollection<Booking> Bookings)
        {
            int count = 0;
            decimal sum = 0;
            foreach (var booking in Bookings)
            {
                if (booking.Rating.HasValue)
                {
                    count++;
                    sum += (int)booking.Rating;
                }
            }
            if (count == 0) return 0;
            return sum / count;
        }
    }
}

File: PawNClaw.Staff.Mobile/lib/models/pet_center.dart
// import 'package:pncstaff_mobile_application/models/photo.dart';

class PetCenter {
  int? id;
  String? name;
  String? address;
  String? phone;
  double? rating;
  String? createDate;
  String? modifyDate;
  int? createUser;
  int? modifyUser;
  bool? status;
  int? brandId;
  String? openTime;
  String? closeTime;
  String? description;
  String? checkin;
  String? checkout;
  int? ratingCount;
  String? endBooking;
  // Photo? photos;

  PetCenter({
    this.id,
    this.name,
    this.address,
    this.phone,
    this.rating,
    this.createDate,
    this.modifyDate,
    this.createUser,
    this.modifyUser,
    this.status,
    this.brandId,
    this.openTime,
    this.closeTime,
    this.description,
    this.checkin,
    this.checkout,
    this.ratingCount,
    this.endBooking,
    //this.photos
  });

  PetCenter.fromJson(Map<String, dynamic> json) {
    id = json['id'];
    name = json['name'];
    address = json['address'];
    phone = json['phone'];
    rating = json['ratingPoint'];
    createDate = json['createDate'];
    modifyDate = json['modifyDate'];
    createUser = json['createUser'];
    modifyUser = json['modifyUser'];
    status = json['status'];
    brandId = json['brandId'];
    openTime = json['openTime'];
    closeTime = json['closeTime'];
    description = json['description'];
    checkin = json['checkin'];
    checkout = json['checkout'];
    ratingCount = json['ratingCount'];
    endBooking = json['endBooking'];

    // photos = Photo.fromJson(json['photos']);
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['id'] = this.id;
    data['name'] = this.name;
    data['address'] = this.address;
    data['phone'] = this.phone;
    data['ratingPoint'] = this.rating!.toDouble();
    data['createDate'] = this.createDate;
    data['modifyDate'] = this.modifyDate;
    data['createUser'] = this.createUser;
    data['modifyUser'] = this.modifyUser;
    data['status'] = this.status;
    data['brandId'] = this.brandId;
    data['openTime'] = this.openTime;
    data['closeTime'] = this.closeTime;
    data['description'] = this.description;
    data['checkin'] = this.checkin;
    data['checkout'] = this.checkout;
    data['ratingCount'] = this.ratingCount;
    data['endBooking'] = this.endBooking;
    // data['photos'] = this.photos?.toJson();
    return data;
  }
}
